crates/stages/stages/src/stages/merkle_changesets.rs

```diff
@@ -0,0 +1,376 @@
+    TrieWriter,
+};
+use reth_stages_api::{
+    BlockErrorKind, CheckpointBlockRange, ExecInput, ExecOutput, MerkleChangeSetsCheckpoint, Stage,
+    StageCheckpoint, StageError, StageId, UnwindInput, UnwindOutput,
+};
+use reth_trie::{updates::TrieUpdates, HashedPostState, KeccakKeyHasher, StateRoot, TrieInput};
+use reth_trie_db::{DatabaseHashedPostState, DatabaseStateRoot};
+use std::ops::Range;
+use tracing::{debug, error};
+
+/// The `MerkleChangeSets` stage.
+///
+/// This stage processes and maintains trie changesets from the finalized block to the latest block.
+#[derive(Debug, Clone)]
+pub struct MerkleChangeSets {
+    /// The number of blocks to retain changesets for, used as a fallback when the finalized block
+    /// is not found. Defaults to 64 (2 epochs in beacon chain).
+    retention_blocks: u64,
+}
+
+impl MerkleChangeSets {
+    /// Creates a new `MerkleChangeSets` stage with default retention blocks of 64.
+    pub const fn new() -> Self {
+        Self { retention_blocks: 64 }
+    }
+
+    /// Creates a new `MerkleChangeSets` stage with a custom finalized block height.
+    pub const fn with_retention_blocks(retention_blocks: u64) -> Self {
+        Self { retention_blocks }
+    }
+
+    /// Returns the range of blocks which are already computed. Will return an empty range if none
+    /// have been computed.
+    fn computed_range(checkpoint: Option<StageCheckpoint>) -> Range<BlockNumber> {
+        let CheckpointBlockRange { from, to } = checkpoint
+            .map(|chk| chk.merkle_changesets_stage_checkpoint().unwrap_or_default())
+            .unwrap_or_default()
+            .block_range;
+        from..to + 1
+    }
+
+    /// Determines the target range for changeset computation based on the checkpoint and provider
+    /// state.
+    ///
+    /// Returns the target range (exclusive end) to compute changesets for.
+    fn determine_target_range<Provider>(
+        &self,
+        provider: &Provider,
+    ) -> Result<Range<BlockNumber>, StageError>
+    where
+        Provider: StageCheckpointReader + ChainStateBlockReader,
+    {
+        // Get merkle checkpoint which represents our target end block
+        let merkle_checkpoint = provider
+            .get_stage_checkpoint(StageId::MerkleExecute)?
+            .map(|checkpoint| checkpoint.block_number)
+            .unwrap_or(0);
+
+        let target_end = merkle_checkpoint + 1; // exclusive
+
+        // Calculate the target range based on the finalized block and the target block.
+        // We maintain changesets from the finalized block to the latest block.
+        let finalized_block = provider.last_finalized_block_number()?;
+
+        // If finalized block is not found, default to being `retention_blocks` away from
+        // target_end
+        let mut target_start = finalized_block
+            .unwrap_or_else(|| merkle_checkpoint.saturating_sub(self.retention_blocks))
```

Comment:
should this instead calculate the target start based on the merkle checkpoint block? Like `merkle_checkpoint - self.retention_blocks`

Reply:
That is what it's doing, if there is no finalized block. I guess a better approach might be to take the min of the two.

Reply:
there's always a finalized block number after the first sync IIUC, so the `retention_blocks` setting will get overwrited by it at all times